multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCache.java

```diff
@@ -0,0 +1,96 @@
+package org.infinispan.multimap.impl;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Predicate;
+
+import org.infinispan.Cache;
+import org.infinispan.commons.api.functional.FunctionalMap;
+import org.infinispan.functional.impl.FunctionalMapImpl;
+import org.infinispan.functional.impl.ReadWriteMapImpl;
+import org.infinispan.multimap.api.MultimapCache;
+import org.infinispan.multimap.impl.function.PutFunction;
+import org.infinispan.multimap.impl.function.RemoveFunction;
+import org.infinispan.util.function.SerializableBiFunction;
+
+/**
+ * Embedded implementation of {@link MultimapCache}
+ *
+ * @author Katia Aresti, karesti@redhat.com
+ * @since 9.1
+ */
+public class EmbeddedMultimapCache<K, V> implements MultimapCache<K, V> {
+
+   private final Cache<K, Collection<V>> cache;
+   private final FunctionalMap.ReadWriteMap<K, Collection<V>> readWriteMap;
+
+   public EmbeddedMultimapCache(Cache<K, Collection<V>> cache) {
+      this.cache = cache;
+      FunctionalMapImpl<K, Collection<V>> functionalMap = FunctionalMapImpl.create(this.cache.getAdvancedCache());
+      this.readWriteMap = ReadWriteMapImpl.create(functionalMap);
+   }
+
+   @Override
+   public CompletableFuture<Void> put(K key, V value) {
+      return readWriteMap.eval(key, new PutFunction<>(key, value));
+   }
+
+   @Override
+   public CompletableFuture<Collection<V>> get(K key) {
+      SerializableBiFunction<? super Collection<V>, Throwable, Collection<V>> emptyCollectionWhenNull = (vs, throwable) -> vs == null ? new HashSet<>() : vs;
+      return cache.getAsync(key).handle(emptyCollectionWhenNull);
```

Comment:
We also need to make sure this is a copy of the collection, just in case. Otherwise if we are in a clustered cache data would get in an inconsistent state if user updates this collection on nodes that own the data.